File: app/frontend/src/app/game/components/2_navigasi_menu/2_navigasi_bawah/5_geopolitik/1_PBB/1_pemungutan_suara/services/aiProposalGenerator.ts
/**
 * AI Proposal Generator
 * Menghasilkan proposal (resolusi, sanksi, embargo) dari negara AI
 * Berdasarkan kepentingan geopolitik, ekonomi, dan stabilitas
 */

import { CountryProfile, Resolution } from './aiVotingService';
import { ProposalType } from '../utils/votingSystem';

export interface AIProposalGeneratorConfig {
  proposalFrequency: number; // Hari antara proposal (default: 5-10 hari)
  minConfidence: number; // Minimum confidence untuk submit proposal (default: 0.6)
  maxActiveProposalsPerCountry: number; // Max proposal aktif per negara (default: 2)
}

export interface GeneratedProposal {
  type: ProposalType;
  proposerCountry: string;
  targetCountry?: string;
  proposalName: string;
  description: string;
  duration: string;
  subItem?: string;
  reasoning: string;
  confidence: number;
  priority: number; // 1-10, semakin tinggi semakin penting
}

class AIProposalGenerator {
  private config: AIProposalGeneratorConfig;

  constructor(config: Partial<AIProposalGeneratorConfig> = {}) {
    this.config = {
      proposalFrequency: config.proposalFrequency || 7,
      minConfidence: config.minConfidence || 0.6,
      maxActiveProposalsPerCountry: config.maxActiveProposalsPerCountry || 2,
    };
  }

  /**
   * Generate proposal untuk satu negara
   */
  generateProposal(
    country: CountryProfile,
    allCountries: CountryProfile[],
    globalTension: number,
    activeProposalsByCountry: number = 0
  ): GeneratedProposal | null {
    // Cek apakah negara sudah punya terlalu banyak proposal aktif
    if (activeProposalsByCountry >= this.config.maxActiveProposalsPerCountry) {
      return null;
    }

    // Tentukan tipe proposal berdasarkan kondisi negara
    const proposalType = this._selectProposalType(country, globalTension);
    if (!proposalType) return null;

    // Tentukan target negara (jika ada)
    const targetCountry = this._selectTargetCountry(country, allCountries, proposalType);

    // Generate proposal details
    const proposal = this._generateProposalDetails(
      country,
      proposalType,
      targetCountry,
      allCountries,
      globalTension
    );

    return proposal;
  }

  /**
   * Tentukan tipe proposal berdasarkan kondisi negara
   */
  private _selectProposalType(
    country: CountryProfile,
    globalTension: number
  ): ProposalType | null {
    const rand = Math.random();

    // Negara dengan stabilitas rendah cenderung propose resolusi
    if (country.stability < 0.4) {
      return rand < 0.7 ? 'resolution' : rand < 0.85 ? 'sanction' : 'embargo';
    }

    // Negara dengan military power tinggi cenderung propose sanksi/embargo
    if (country.military_power > 0.7) {
      return rand < 0.4 ? 'sanction' : rand < 0.7 ? 'embargo' : 'resolution';
    }

    // Negara netral cenderung propose resolusi
    if (country.political_ideology === 'neutral' || country.political_ideology === 'non_aligned') {
      return rand < 0.6 ? 'resolution' : rand < 0.8 ? 'sanction' : 'embargo';
    }

    // Default: random
    return rand < 0.4 ? 'resolution' : rand < 0.7 ? 'sanction' : 'embargo';
  }

  /**
   * Pilih negara target untuk sanksi/embargo
   */
  private _selectTargetCountry(
    country: CountryProfile,
    allCountries: CountryProfile[],
    proposalType: ProposalType
  ): CountryProfile | undefined {
    if (proposalType === 'resolution') {
      return undefined; // Resolusi tidak punya target
    }

    // Cari negara dengan hubungan buruk atau ideologi berbeda
    const candidates = allCountries.filter(c => {
      if (c.name === country.name) return false;

      // Prioritas: ideologi berbeda, military power tinggi, atau stability rendah
      const ideologyDiff = country.political_ideology !== c.political_ideology ? 1 : 0;
      const militaryThreat = c.military_power > 0.6 ? 1 : 0;
      const stabilityRisk = c.stability < 0.4 ? 1 : 0;

      return ideologyDiff + militaryThreat + stabilityRisk > 0;
    });

    if (candidates.length === 0) return undefined;

    // Pilih random dari candidates
    return candidates[Math.floor(Math.random() * candidates.length)];
  }

  /**
   * Generate detail proposal
   */
  private _generateProposalDetails(
    country: CountryProfile,
    type: ProposalType,
    targetCountry: CountryProfile | undefined,
    allCountries: CountryProfile[],
    globalTension: number
  ): GeneratedProposal {
    const templates = this._getProposalTemplates(type, country, targetCountry);
    const template = templates[Math.floor(Math.random() * templates.length)];

    // Hitung confidence berdasarkan faktor
    const confidence = this._calculateProposalConfidence(
      country,
      type,
      targetCountry,
      allCountries
    );

    // Hitung priority
    const priority = this._calculatePriority(country, type, globalTension, confidence);

    return {
      type,
      proposerCountry: country.name,
      targetCountry: targetCountry?.name,
      proposalName: template.name,
      description: template.description,
      duration: template.duration,
      subItem: template.subItem,
      reasoning: template.reasoning,
      confidence,
      priority,
    };
  }

  /**
   * Template proposal untuk berbagai skenario
   */
  private _getProposalTemplates(
    type: ProposalType,
    country: CountryProfile,
    targetCountry?: CountryProfile
  ) {
    if (type === 'resolution') {
      return [
        {
          name: 'Resolusi Perdamaian Global',
          description: 'Mendorong dialog internasional dan penyelesaian konflik secara damai',
          duration: '30 hari',
          reasoning: 'Stabilitas global diperlukan untuk pertumbuhan ekonomi',
        },
        {
          name: 'Resolusi Perlindungan Lingkungan',
          description: 'Komitmen bersama untuk mengurangi emisi karbon dan melindungi ekosistem',
          duration: '60 hari',
          reasoning: 'Perubahan iklim mempengaruhi semua negara',
        },
        {
          name: 'Resolusi Kerjasama Ekonomi',
          description: 'Meningkatkan perdagangan dan investasi antar negara',
          duration: '30 hari',
          reasoning: 'Pertumbuhan ekonomi bersama menguntungkan semua pihak',
        },
      ];
    }

    if (type === 'sanction') {
      const targetName = targetCountry?.name || 'Negara Target';
      return [
        {
          name: `Sanksi Ekonomi terhadap ${targetName}`,
          description: `Membatasi perdagangan dan investasi dengan ${targetName} karena pelanggaran hak asasi manusia`,
          duration: '90 hari',
          reasoning: 'Tekanan ekonomi untuk perubahan kebijakan',
        },
        {
          name: `Embargo Senjata terhadap ${targetName}`,
          description: `Melarang penjualan senjata dan teknologi militer ke ${targetName}`,
          duration: '180 hari',
          reasoning: 'Mencegah eskalasi konflik regional',
        },
      ];
    }

    // embargo
    const targetName = targetCountry?.name || 'Negara Target';
    return [
      {
        name: `Embargo Minyak terhadap ${targetName}`,
        description: `Melarang ekspor minyak dan produk energi dari ${targetName}`,
        duration: '120 hari',
        subItem: 'energi',
        reasoning: 'Tekanan untuk mengubah kebijakan luar negeri',
      },
      {
        name: `Embargo Teknologi terhadap ${targetName}`,
        description: `Melarang transfer teknologi dan akses ke pasar teknologi global`,
        duration: '180 hari',
        subItem: 'teknologi',
        reasoning: 'Membatasi kemampuan militer dan ekonomi',
      },
    ];
  }

  /**
   * Hitung confidence proposal
   */
  private _calculateProposalConfidence(
    country: CountryProfile,
    type: ProposalType,
    targetCountry: CountryProfile | undefined,
    allCountries: CountryProfile[]
  ): number {
    let confidence = 0.5;

    // Faktor 1: Diplomatic standing
    confidence += (country.diplomatic_standing / 100) * 0.2;

    // Faktor 2: Ideologi alignment dengan negara lain
    const alignedCountries = allCountries.filter(
      c => c.political_ideology === country.political_ideology && c.name !== country.name
    ).length;
    confidence += (alignedCountries / allCountries.length) * 0.2;

    // Faktor 3: Untuk sanction/embargo, cek apakah target adalah musuh
    if (type !== 'resolution' && targetCountry) {
      const isEnemy = targetCountry.political_ideology !== country.political_ideology;
      confidence += isEnemy ? 0.15 : -0.1;
    }

    // Faktor 4: Military power (untuk sanction/embargo)
    if (type !== 'resolution') {
      confidence += (country.military_power / 2) * 0.15;
    }

    return Math.max(0.3, Math.min(1.0, confidence));
  }

  /**
   * Hitung priority proposal
   */
  private _calculatePriority(
    country: CountryProfile,
    type: ProposalType,
    globalTension: number,
    confidence: number
  ): number {
    let priority = 5; // Base priority

    // Tinggi jika confidence tinggi
    priority += confidence * 3;

    // Tinggi jika global tension tinggi dan type adalah resolution
    if (type === 'resolution' && globalTension > 0.7) {
      priority += 2;
    }

    // Tinggi jika negara punya military power tinggi dan type adalah sanction/embargo
    if (type !== 'resolution' && country.military_power > 0.7) {
      priority += 2;
    }

    return Math.min(10, Math.max(1, Math.round(priority)));
  }
}

export const aiProposalGenerator = new AIProposalGenerator();

File: app/frontend/src/app/game/components/2_navigasi_menu/2_navigasi_bawah/5_geopolitik/1_PBB/1_pemungutan_suara/0_pemungutan_AI_logic/ai_voting_service/ai_proposal_generator.py
"""
AI Proposal Generator Service
Menghasilkan proposal dari negara AI berdasarkan kepentingan geopolitik
"""

from typing import List, Optional, Literal
from pydantic import BaseModel
from enum import Enum
import random
import json

# ==================== MODELS ====================

class ProposalType(str, Enum):
    RESOLUTION = "resolution"
    SANCTION = "sanction"
    EMBARGO = "embargo"

class CountryProfile(BaseModel):
    name: str
    continent: str
    gdp: float
    military_power: float
    diplomatic_standing: float
    alliance_bloc: Optional[str] = None
    economic_dependency: dict = {}
    political_ideology: str = "neutral"  # "western", "eastern", "neutral", "non_aligned"
    stability: float = 0.7

class GeneratedProposal(BaseModel):
    type: ProposalType
    proposer_country: str
    target_country: Optional[str] = None
    proposal_name: str
    description: str
    duration: str
    sub_item: Optional[str] = None
    reasoning: str
    confidence: float
    priority: int  # 1-10

class ProposalGenerationRequest(BaseModel):
    country: CountryProfile
    all_countries: List[CountryProfile]
    global_tension: float = 0.5
    active_proposals_count: int = 0

class ProposalGenerationResponse(BaseModel):
    proposal: Optional[GeneratedProposal] = None
    reason: str = ""

# ==================== PROPOSAL GENERATOR ====================

class AIProposalGenerator:
    """
    Menghasilkan proposal untuk negara AI
    """
    
    def __init__(self):
        self.max_active_proposals_per_country = 2
        self.min_confidence = 0.6
    
    def generate_proposal(
        self,
        country: CountryProfile,
        all_countries: List[CountryProfile],
        global_tension: float,
        active_proposals_count: int = 0
    ) -> Optional[GeneratedProposal]:
        """
        Generate proposal untuk satu negara
        """
        
        # Cek apakah negara sudah punya terlalu banyak proposal aktif
        if active_proposals_count >= self.max_active_proposals_per_country:
            return None
        
        # Tentukan tipe proposal
        proposal_type = self._select_proposal_type(country, global_tension)
        if not proposal_type:
            return None
        
        # Tentukan target negara (jika ada)
        target_country = self._select_target_country(country, all_countries, proposal_type)
        
        # Generate proposal details
        proposal = self._generate_proposal_details(
            country,
            proposal_type,
            target_country,
            all_countries,
            global_tension
        )
        
        return proposal
    
    def _select_proposal_type(
        self,
        country: CountryProfile,
        global_tension: float
    ) -> Optional[ProposalType]:
        """
        Tentukan tipe proposal berdasarkan kondisi negara
        """
        rand = random.random()
        
        # Negara dengan stabilitas rendah cenderung propose resolusi
        if country.stability < 0.4:
            return ProposalType.RESOLUTION if rand < 0.7 else (
                ProposalType.SANCTION if rand < 0.85 else ProposalType.EMBARGO
            )
        
        # Negara dengan military power tinggi cenderung propose sanksi/embargo
        if country.military_power > 0.7:
            return ProposalType.SANCTION if rand < 0.4 else (
                ProposalType.EMBARGO if rand < 0.7 else ProposalType.RESOLUTION
            )
        
        # Negara netral cenderung propose resolusi
        if country.political_ideology in ["neutral", "non_aligned"]:
            return ProposalType.RESOLUTION if rand < 0.6 else (
                ProposalType.SANCTION if rand < 0.8 else ProposalType.EMBARGO
            )
        
        # Default: random
        rand_type = random.random()
        if rand_type < 0.4:
            return ProposalType.RESOLUTION
        elif rand_type < 0.7:
            return ProposalType.SANCTION
        else:
            return ProposalType.EMBARGO
    
    def _select_target_country(
        self,
        country: CountryProfile,
        all_countries: List[CountryProfile],
        proposal_type: ProposalType
    ) -> Optional[CountryProfile]:
        """
        Pilih negara target untuk sanksi/embargo
        """
        if proposal_type == ProposalType.RESOLUTION:
            return None
        
        # Cari negara dengan hubungan buruk atau ideologi berbeda
        candidates = []
        for c in all_countries:
            if c.name == country.name:
                continue
            
            # Prioritas: ideologi berbeda, military power tinggi, atau stability rendah
            ideology_diff = 1 if country.political_ideology != c.political_ideology else 0
            military_threat = 1 if c.military_power > 0.6 else 0
            stability_risk = 1 if c.stability < 0.4 else 0
            
            score = ideology_diff + military_threat + stability_risk
            if score > 0:
                candidates.append((c, score))
        
        if not candidates:
            return None
        
        # Pilih random dari candidates (weighted by score)
        total_score = sum(score for _, score in candidates)
        rand_val = random.uniform(0, total_score)
        current = 0
        
        for candidate, score in candidates:
            current += score
            if rand_val <= current:
                return candidate
        
        return candidates[-1][0] if candidates else None
    
    def _generate_proposal_details(
        self,
        country: CountryProfile,
        proposal_type: ProposalType,
        target_country: Optional[CountryProfile],
        all_countries: List[CountryProfile],
        global_tension: float
    ) -> GeneratedProposal:
        """
        Generate detail proposal
        """
        templates = self._get_proposal_templates(proposal_type, country, target_country)
        template = random.choice(templates)
        
        # Hitung confidence
        confidence = self._calculate_confidence(
            country,
            proposal_type,
            target_country,
            all_countries
        )
        
        # Hitung priority
        priority = self._calculate_priority(country, proposal_type, global_tension, confidence)
        
        return GeneratedProposal(
            type=proposal_type,
            proposer_country=country.name,
            target_country=target_country.name if target_country else None,
            proposal_name=template["name"],
            description=template["description"],
            duration=template["duration"],
            sub_item=template.get("sub_item"),
            reasoning=template["reasoning"],
            confidence=confidence,
            priority=priority
        )
    
    def _get_proposal_templates(
        self,
        proposal_type: ProposalType,
        country: CountryProfile,
        target_country: Optional[CountryProfile]
    ) -> List[dict]:
        """
        Template proposal untuk berbagai skenario
        """
        if proposal_type == ProposalType.RESOLUTION:
            return [
                {
                    "name": "Resolusi Perdamaian Global",
                    "description": "Mendorong dialog internasional dan penyelesaian konflik secara damai",
                    "duration": "30 hari",
                    "reasoning": "Stabilitas global diperlukan untuk pertumbuhan ekonomi"
                },
                {
                    "name": "Resolusi Perlindungan Lingkungan",
                    "description": "Komitmen bersama untuk mengurangi emisi karbon dan melindungi ekosistem",
                    "duration": "60 hari",
                    "reasoning": "Perubahan iklim mempengaruhi semua negara"
                },
                {
                    "name": "Resolusi Kerjasama Ekonomi",
                    "description": "Meningkatkan perdagangan dan investasi antar negara",
                    "duration": "30 hari",
                    "reasoning": "Pertumbuhan ekonomi bersama menguntungkan semua pihak"
                },
                {
                    "name": "Resolusi Hak Asasi Manusia",
                    "description": "Perlindungan universal hak asasi manusia dan kebebasan fundamental",
                    "duration": "45 hari",
                    "reasoning": "Hak asasi manusia adalah fondasi perdamaian dunia"
                }
            ]
        
        if proposal_type == ProposalType.SANCTION:
            target_name = target_country.name if target_country else "Negara Target"
            return [
                {
                    "name": f"Sanksi Ekonomi terhadap {target_name}",
                    "description": f"Membatasi perdagangan dan investasi dengan {target_name} karena pelanggaran hak asasi manusia",
                    "duration": "90 hari",
                    "reasoning": "Tekanan ekonomi untuk perubahan kebijakan"
                },
                {
                    "name": f"Embargo Senjata terhadap {target_name}",
                    "description": f"Melarang penjualan senjata dan teknologi militer ke {target_name}",
                    "duration": "180 hari",
                    "reasoning": "Mencegah eskalasi konflik regional"
                },
                {
                    "name": f"Pembekuan Aset terhadap {target_name}",
                    "description": f"Membekukan aset dan akun keuangan {target_name} di lembaga internasional",
                    "duration": "120 hari",
                    "reasoning": "Tekanan finansial untuk compliance dengan resolusi PBB"
                }
            ]
        
        # EMBARGO
        target_name = target_country.name if target_country else "Negara Target"
        return [
            {
                "name": f"Embargo Minyak terhadap {target_name}",
                "description": f"Melarang ekspor minyak dan produk energi dari {target_name}",
                "duration": "120 hari",
                "sub_item": "energi",
                "reasoning": "Tekanan untuk mengubah kebijakan luar negeri"
            },
            {
                "name": f"Embargo Teknologi terhadap {target_name}",
                "description": f"Melarang transfer teknologi dan akses ke pasar teknologi global",
                "duration": "180 hari",
                "sub_item": "teknologi",
                "reasoning": "Membatasi kemampuan militer dan ekonomi"
            },
            {
                "name": f"Embargo Pertanian terhadap {target_name}",
                "description": f"Melarang ekspor produk pertanian dan pangan dari {target_name}",
                "duration": "90 hari",
                "sub_item": "pertanian",
                "reasoning": "Tekanan ekonomi untuk perubahan kebijakan"
            }
        ]
    
    def _calculate_confidence(
        self,
        country: CountryProfile,
        proposal_type: ProposalType,
        target_country: Optional[CountryProfile],
        all_countries: List[CountryProfile]
    ) -> float:
        """
        Hitung confidence proposal
        """
        confidence = 0.5
        
        # Faktor 1: Diplomatic standing
        confidence += (country.diplomatic_standing / 100) * 0.2
        
        # Faktor 2: Ideologi alignment dengan negara lain
        aligned_countries = sum(
            1 for c in all_countries
            if c.political_ideology == country.political_ideology and c.name != country.name
        )
        confidence += (aligned_countries / len(all_countries)) * 0.2
        
        # Faktor 3: Untuk sanction/embargo, cek apakah target adalah musuh
        if proposal_type != ProposalType.RESOLUTION and target_country:
            is_enemy = target_country.political_ideology != country.political_ideology
            confidence += 0.15 if is_enemy else -0.1
        
        # Faktor 4: Military power (untuk sanction/embargo)
        if proposal_type != ProposalType.RESOLUTION:
            confidence += (country.military_power / 2) * 0.15
        
        return max(0.3, min(1.0, confidence))
    
    def _calculate_priority(
        self,
        country: CountryProfile,
        proposal_type: ProposalType,
        global_tension: float,
        confidence: float
    ) -> int:
        """
        Hitung priority proposal (1-10)
        """
        priority = 5
        
        # Tinggi jika confidence tinggi
        priority += confidence * 3
        
        # Tinggi jika global tension tinggi dan type adalah resolution
        if proposal_type == ProposalType.RESOLUTION and global_tension > 0.7:
            priority += 2
        
        # Tinggi jika negara punya military power tinggi dan type adalah sanction/embargo
        if proposal_type != ProposalType.RESOLUTION and country.military_power > 0.7:
            priority += 2
        
        return min(10, max(1, round(priority)))


# ==================== API ENDPOINTS ====================

from fastapi import APIRouter

router = APIRouter(prefix="/proposal", tags=["proposal"])
generator = AIProposalGenerator()

@router.post("/generate")
async def generate_proposal(request: ProposalGenerationRequest) -> ProposalGenerationResponse:
    """
    Generate proposal untuk satu negara
    """
    try:
        proposal = generator.generate_proposal(
            request.country,
            request.all_countries,
            request.global_tension,
            request.active_proposals_count
        )
        
        if proposal:
            return ProposalGenerationResponse(
                proposal=proposal,
                reason="Proposal generated successfully"
            )
        else:
            return ProposalGenerationResponse(
                proposal=None,
                reason="Country already has max active proposals or confidence too low"
            )
    
    except Exception as e:
        return ProposalGenerationResponse(
            proposal=None,
            reason=f"Error generating proposal: {str(e)}"
        )

@router.post("/generate-batch")
async def generate_batch_proposals(
    countries: List[CountryProfile],
    global_tension: float = 0.5,
    active_proposals_per_country: dict = {}
) -> dict:
    """
    Generate proposals untuk multiple negara
    """
    try:
        proposals = []
        
        for country in countries:
            if country.name == "Player":
                continue
            
            active_count = active_proposals_per_country.get(country.name, 0)
            proposal = generator.generate_proposal(
                country,
                countries,
                global_tension,
                active_count
            )
            
            if proposal:
                proposals.append(proposal)
        
        return {
            "success": True,
            "proposals": proposals,
            "count": len(proposals)
        }
    
    except Exception as e:
        return {
            "success": False,
            "error": str(e),
            "proposals": [],
            "count": 0
        }
